Run on Elixir 1.15 and Erlang 26

dabaer opened the pull request:
To be merged at the same time as

- https://github.com/exercism/elixir-analyzer/pull/386
- https://github.com/exercism/elixir-representer/pull/84

dabaer commented on Dockerfile:
```diff
@@ -1,8 +1,8 @@
-FROM hexpm/elixir:1.14.0-erlang-25.0.4-debian-bullseye-20220801
```
There doesn't seem to be a debian based docker image for 1.15/26, but I see ubuntu has been used in prior version updates here

dabaer replied:
Ah I was looking at plain 26.0, I'll update to 26.0.1 and use the debian images

neenjaw replied:
debian buster is older than debian bullseye.  If you could use [1.15.0-erlang-26.0.1-debian-bookworm-20230612](https://hub.docker.com/layers/hexpm/elixir/1.15.0-erlang-26.0.1-debian-bookworm-20230612/images/sha256-025bc49b88562025fac1479e338592e469670076bcbc43542d5e8e9ca144f31d?context=explore) that would be a preferable option.

dabaer replied:
Good idea, I will update the PRs

dabaer commented:
@angelikatyborska I'm currently running into an error related to the `results.json` file, but I can't quite find at what point in the process this file is actually supposed to be created. I thought the `exercism_test_helper` initially but I don't see any references to it's creation in `run.sh` or in the helper source. Which module is creating it?

```elixir
** (File.Error) could not read file "/tmp/solution_VMpMp412od/results.json": no such file or directory
    (elixir 1.15.0) lib/file.ex:358: File.read!/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:7: anonymous fn/1 in ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:11: ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (elixir 1.15.0) lib/kernel/cli.ex:136: anonymous fn/3 in Kernel.CLI.exec_fun/2
cp: cannot stat '/tmp/solution_VMpMp412od/results.json': No such file or directory
```

angelikatyborska commented:
Sorry about the autoclosing github action, it's an oversight that it was added to this repo.

angelikatyborska commented:
Another thing I forgot about - https://github.com/exercism/elixir-test-runner/pull/107 needs to be merged to allow CI to pass on any other PR 🤦

angelikatyborska commented:
> I thought the exercism_test_helper initially but I don't see any references to it's creation in run.sh or in the helper source. Which module is creating it?

If I remember correctly, the file gets created here: 

https://github.com/exercism/elixir-test-runner/blob/f6d125e8c30d33d70a1a99b5b0bfe6b95cfbde64/exercism_test_helper/lib/json_formatter.ex#L56-L63

dabaer commented:
@angelikatyborska thusfar i've been following the codepaths running inside of docker and discovered that everything is fine until Elixir 1.14.3, at which point the changes to `ExUnit.Server.modules_loaded()` where an argument was now required. That was fine, it was one of the first things I fixed when running 1.15 builds of docker.

Things got a little interesting with 14.4, tests were failing due to changes in the stacktrace line numbers changing by 1. This only happens on the `community_garden` test for some reason, and I'm not sure what the root cause of that is.

```
parsing /tmp/solution_ghuF42hNhM/test/community_garden_test.exs for metadata
transforming /tmp/solution_ghuF42hNhM/test/community_garden_test.exs
warning: variable "pid" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:14: CommunityGarden.list_registrations/1

warning: variable "pid" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:18: CommunityGarden.register/2

warning: variable "register_to" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:18: CommunityGarden.register/2

warning: variable "pid" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:22: CommunityGarden.release/2

warning: variable "plot_id" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:22: CommunityGarden.release/2

warning: variable "pid" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/community_garden.ex:26: CommunityGarden.get_registration/2

warning: variable "plot_id" is unused (if the variable is not meant to be used, prefix it with an underscore)

  lib/community_garden.ex:26: CommunityGarden.get_registration/2

22c22
<       "message": "  2) test can register a new plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:18\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:22: (test)\n",
---
>       "message": "  2) test can register a new plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:18\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:21: (test)\n",
30c30
<       "message": "  3) test maintains a registry of plots (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:30\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:34: (test)\n",
---
>       "message": "  3) test maintains a registry of plots (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:25\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:28: (test)\n",
38c38
<       "message": "  4) test can release a plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:45\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:49: (test)\n",
---
>       "message": "  4) test can release a plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:33\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:36: (test)\n",
46c46
<       "message": "  5) test can get registration of a registered plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:61\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:65: (test)\n",
---
>       "message": "  5) test can get registration of a registered plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:42\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert %Plot{} = plot = CommunityGarden.register(pid, \"Johnny Appleseed\")\n     \u001b[36mleft:  \u001b[0m\u001b[31m%Plot{\n  \n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:45: (test)\n",
54c54
<       "message": "  6) test return not_found when attempt to get registration of an unregistered plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:78\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert {:not_found, \"plot is unregistered\"} = CommunityGarden.get_registration(pid, 1)\n     \u001b[36mleft:  \u001b[0m\u001b[31m{\n  :not_found,\n  \"plot is unregistered\"\n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:86: (test)\n",
---
>       "message": "  6) test return not_found when attempt to get registration of an unregistered plot (CommunityGardenTest)\n     \u001b[1m\u001b[30mtest/community_garden_test.exs:52\u001b[0m\n     \u001b[31mmatch (=) failed\u001b[0m\n     \u001b[36mcode:  \u001b[0massert {:not_found, \"plot is unregistered\"} = CommunityGarden.get_registration(pid, 1)\n     \u001b[36mleft:  \u001b[0m\u001b[31m{\n  :not_found,\n  \"plot is unregistered\"\n}\u001b[0m\n     \u001b[36mright: \u001b[0m\u001b[32mnil\u001b[0m\n     \u001b[36mstacktrace:\u001b[0m\n       test/community_garden_test.exs:59: (test)\n",
🔥 test/community-garden: expected /tmp/output_mLCWzQxISL/results.json to equal test/community-garden/expected_results.json on successful run 🔥
```

In the interest of finding my main issues with 1.15 I adjusted `expected_results.json` to match the new output.

Fantastic, now we are running fine on 1.14.5. Incrementing bit by bit, switching to an Erlang 26 docker release, we are still smooth sailing. Even better!

1.15.0 is where we run into a brick wall.

```
parsing /tmp/solution_PY6PXRW4kn/test/rpg_test.exs for metadata
transforming /tmp/solution_PY6PXRW4kn/test/rpg_test.exs
warning: redefining module RPG (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.beam)
  lib/rpg.ex:47: RPG (module)

warning: redefining module RPG.Character (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Character.beam)
  lib/rpg.ex:48: RPG.Character (module)

warning: redefining module RPG.LoafOfBread (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.LoafOfBread.beam)
  lib/rpg.ex:52: RPG.LoafOfBread (module)

warning: redefining module RPG.ManaPotion (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.ManaPotion.beam)
  lib/rpg.ex:56: RPG.ManaPotion (module)

warning: redefining module RPG.Poison (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Poison.beam)
  lib/rpg.ex:60: RPG.Poison (module)

warning: redefining module RPG.EmptyBottle (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.EmptyBottle.beam)
  lib/rpg.ex:64: RPG.EmptyBottle (module)

warning: redefining module RPG.Edible (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.beam)
  lib/rpg.ex:68: RPG.Edible (module)

warning: redefining module RPG.Edible.RPG.LoafOfBread (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.LoafOfBread.beam)
  lib/rpg.ex:72: RPG.Edible.RPG.LoafOfBread (module)

warning: redefining module RPG.Edible.RPG.ManaPotion (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.ManaPotion.beam)
  lib/rpg.ex:79: RPG.Edible.RPG.ManaPotion (module)

warning: redefining module RPG.Edible.RPG.Poison (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.Poison.beam)
  lib/rpg.ex:86: RPG.Edible.RPG.Poison (module)

** (File.Error) could not read file "/tmp/solution_PY6PXRW4kn/results.json": no such file or directory
    (elixir 1.15.0) lib/file.ex:358: File.read!/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:7: anonymous fn/1 in ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:11: ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (elixir 1.15.0) lib/kernel/cli.ex:136: anonymous fn/3 in Kernel.CLI.exec_fun/2
cp: cannot stat '/tmp/solution_PY6PXRW4kn/results.json': No such file or directory
```

I poked into it based on your direction with `json_formatter.ex`, and for some reason `handle_cast` for `:suite_finished` isn't getting called at all. This leads me to believe the tests simply aren't running at all, and not producing a `results.json`.

At first I reneged on changing `modules_loaded()` to `modules_loaded(false)` and dug around the source code for `ExUnit.Server`, but this didn't really lead anywhere. `true` and `false` result in the same issue. Frustratingly, there doesn't seem to be any documentation on this change other than a commit from Jose `ExUnit improvements`. `ExUnit.Server` doesn't seem to have any documentation on hex (maybe it's a private interface?).

Rats. Well I figured the best thing to do was to was to break out of docker and start running things on bare metal (okay, fine a VM). Another brick wall! Running on a fresh clone of `main` from the Exercism repo on Elixir 1.14.0 and Erlang 25.0.4, I get a suspiciously similar error as when we upgrade to 1.15.

```
dakora@ubuntu-vm:~$ git clone https://github.com/exercism/elixir-test-runner.git
Cloning into 'elixir-test-runner'...
vremote: Enumerating objects: 799, done.
remote: Counting objects: 100% (208/208), done.
remote: Compressing objects: 100% (120/120), done.
remote: Total 799 (delta 149), reused 109 (delta 87), pack-reused 591
Receiving objects: 100% (799/799), 170.84 KiB | 452.00 KiB/s, done.
Resolving deltas: 100% (429/429), done.
dakora@ubuntu-vm:~$ cd elixir-test-runner/
dakora@ubuntu-vm:~/elixir-test-runner$ cd exercism_test_helper/
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ mix local.rebar --force
* creating /home/dakora/.asdf/installs/elixir/1.14.0-otp-25/.mix/elixir/1-14/rebar3
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ mix local.hex --force
* creating /home/dakora/.asdf/installs/elixir/1.14.0-otp-25/.mix/archives/hex-2.0.6
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ mix deps.get
Resolving Hex dependencies...
Resolution completed in 0.026s
Unchanged:
  jason 1.2.2
* Getting jason (Hex package)
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ MIX_ENV=test mix compile
==> jason
Compiling 8 files (.ex)
Generated jason app
==> exercism_test_helper
Compiling 12 files (.ex)
Generated exercism_test_helper app
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ mix test --no-compile
...................................
Finished in 0.2 seconds (0.00s async, 0.2s sync)
35 tests, 0 failures

Randomized with seed 0
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ MIX_ENV=prod mix escript.build
==> jason
Compiling 8 files (.ex)
Generated jason app
==> exercism_test_helper
Compiling 12 files (.ex)
Generated exercism_test_helper app
Generated escript exercism_test_helper with MIX_ENV=prod
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ mv exercism_test_helper ../bin/
dakora@ubuntu-vm:~/elixir-test-runner/exercism_test_helper$ cd ..
dakora@ubuntu-vm:~/elixir-test-runner$ ./bin/smoke_test.sh
parsing /tmp/solution_tLj2UoGmNc/test/rpg_test.exs for metadata
transforming /tmp/solution_tLj2UoGmNc/test/rpg_test.exs
warning: redefining module RPG (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.beam)
  lib/rpg.ex:47: RPG (module)

warning: redefining module RPG.Character (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Character.beam)
  lib/rpg.ex:48: RPG.Character (module)

warning: redefining module RPG.LoafOfBread (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.LoafOfBread.beam)
  lib/rpg.ex:52: RPG.LoafOfBread (module)

warning: redefining module RPG.ManaPotion (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.ManaPotion.beam)
  lib/rpg.ex:56: RPG.ManaPotion (module)

warning: redefining module RPG.Poison (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Poison.beam)
  lib/rpg.ex:60: RPG.Poison (module)

warning: redefining module RPG.EmptyBottle (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.EmptyBottle.beam)
  lib/rpg.ex:64: RPG.EmptyBottle (module)

warning: redefining module RPG.Edible (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.beam)
  lib/rpg.ex:68: RPG.Edible (module)

warning: redefining module RPG.Edible.RPG.LoafOfBread (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.LoafOfBread.beam)
  lib/rpg.ex:72: RPG.Edible.RPG.LoafOfBread (module)

warning: redefining module RPG.Edible.RPG.ManaPotion (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.ManaPotion.beam)
  lib/rpg.ex:79: RPG.Edible.RPG.ManaPotion (module)

warning: redefining module RPG.Edible.RPG.Poison (current version loaded from _build/test/lib/bread_and_potions/ebin/Elixir.RPG.Edible.RPG.Poison.beam)
  lib/rpg.ex:86: RPG.Edible.RPG.Poison (module)

** (File.Error) could not read file "/tmp/solution_tLj2UoGmNc/results.json": no such file or directory
    (elixir 1.14.0) lib/file.ex:358: File.read!/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:7: anonymous fn/1 in ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (exercism_test_helper 0.1.2) lib/exercism_test_helper/cli/command/combine_json.ex:11: ExercismTestHelper.CLI.Command.CombineJSON.run/1
    (elixir 1.14.0) lib/kernel/cli.ex:131: anonymous fn/3 in Kernel.CLI.exec_fun/2
cp: cannot stat '/tmp/solution_tLj2UoGmNc/results.json': No such file or directory
```

There seems to be something going awry in the `test_transform` portion that fails silently. Any guidance would be appreciated, especially in the realm of running this in not-docker. I'm not sure if I'm setting this up correctly, maybe it's something weird when running on Ubuntu.

My next course of action will currently be trying to run on a Debian VM, in order to get a little closer to the Docker environment, but that will be a task for tomorrow!

angelikatyborska commented:
> Any guidance would be appreciated, especially in the realm of running this in not-docker. I'm not sure if I'm setting this up correctly, maybe it's something weird when running on Ubuntu.

I am also not that knowledgable about the set up of this project. What I usually do to run it locally is:
- `cd exercism_test_helper`
- `MIX_ENV=prod mix escript.build`
- `rm ../bin/exercism_test_helper`
- `mv exercism_test_helper ../bin`
- `cd ..`
- `./bin/smoke_test.sh`

To adjust which smoke test runs, I just edit the glob in `for solution in test/*`.

Anyway, I poked around on my own. Based on what you wrote, I had a hunch that the JSONFormatter doesn't get used, or crashes on start, or something like that. I added a `test/community-garden/config/config.exs` file like this to give me more visibility into what happens:
```
import Config

config :logger,
       handle_otp_reports: true,
       handle_sasl_reports: true
```

And I got this:

```
18:42:25.945 [info] Application logger started at :nonode@nohost

18:42:25.947 [info] Application community_garden started at :nonode@nohost

18:42:25.952 [info] Child ExUnit.Server of Supervisor ExUnit.Supervisor started
Pid: #PID<0.130.0>
Start Call: ExUnit.Server.start_link([])
Restart: :permanent
Shutdown: 5000
Type: :worker

18:42:25.952 [info] Child ExUnit.CaptureServer of Supervisor ExUnit.Supervisor started
Pid: #PID<0.131.0>
Start Call: ExUnit.CaptureServer.start_link([])
Restart: :permanent
Shutdown: 5000
Type: :worker

18:42:25.952 [info] Child ExUnit.OnExitHandler of Supervisor ExUnit.Supervisor started
Pid: #PID<0.132.0>
Start Call: ExUnit.OnExitHandler.start_link([])
Restart: :permanent
Shutdown: 5000
Type: :worker

18:42:25.952 [info] Application ex_unit started at :nonode@nohost

18:42:25.965 [error] Process #PID<0.140.0> terminating
** (UndefinedFunctionError) function JSONFormatter.init/1 is undefined (module JSONFormatter is not available)
    JSONFormatter.init([include: [], exclude: [:slow, :pending], max_cases: 1, max_failures: :infinity, failures_manifest_file: "/private/tmp/solution_cyFCunpW8B/_build/test/lib/community_garden/.mix/.mix_test_failures", seed: 0, autorun: false, stacktrace_depth: 20, slowest: 0, formatters: [JSONFormatter], colors: [], exit_status: 2, refute_receive_timeout: 100, capture_log: false, after_suite: [], assert_receive_timeout: 100, trace: true, timeout: 60000])
    (stdlib 5.0.1) gen_server.erl:962: :gen_server.init_it/2
    (stdlib 5.0.1) gen_server.erl:917: :gen_server.init_it/6
    (stdlib 5.0.1) proc_lib.erl:241: :proc_lib.init_p_do_apply/3
Initial Call: JSONFormatter.init/1
Ancestors: [#PID<0.137.0>, #PID<0.133.0>, #PID<0.98.0>]
Message Queue Length: 0
Messages: []
Links: [#PID<0.137.0>]
Dictionary: []
Trapping Exits: false
Status: :running
Heap Size: 376
Stack Size: 28
Reductions: 220
```

@neenjaw do you have any idea - [in this elixir command](https://github.com/exercism/elixir-test-runner/blob/f9f4c5c57a45bb076ea857e115e81eef9b2e645b/bin/run.sh#L59-L67), how is the VM supposed to know what the `JSONFormatter` module is? How does it get loaded? Via this flag: `-pa ${base_dir}/exercism_test_helper/_build/test/lib/exercism_test_helper/ebin`? It seems like that doesn't work anymore.

dabaer commented:
> > Any guidance would be appreciated, especially in the realm of running this in not-docker. I'm not sure if I'm setting this up correctly, maybe it's something weird when running on Ubuntu.
> 
> I am also not that knowledgable about the set up of this project. What I usually do to run it locally is:
> 
> * `cd exercism_test_helper`
> * `MIX_ENV=prod mix escript.build`
> * `rm ../bin/exercism_test_helper`
> * `mv exercism_test_helper ../bin`
> * `cd ..`
> * `./bin/smoke_test.sh`
> 
> To adjust which smoke test runs, I just edit the glob in `for solution in test/*`.

Okay perfect, I was worried I was missing something to do with setup!

> Anyway, I poked around on my own. Based on what you wrote, I had a hunch that the JSONFormatter doesn't get used, or crashes on start, or something like that. I added a `test/community-garden/config/config.exs` file like this to give me more visibility into what happens:
> 
> ```
> import Config
> 
> config :logger,
>        handle_otp_reports: true,
>        handle_sasl_reports: true
> ```
> 
> And I got this:

That's interesting, though do you also get that error when running on the stock repo with 1.14, or just 1.15? That's where I got stuck on back-to-stock-repo sanity checks. I was really thinking it was my environment

dabaer commented:
I think you've hit nail on the head, it appears to be a code path trimming issue.

Initially, I had tried this by adding `prune_code_paths: false` to the `mix.exs` for `exercism_test_helper`, but I didn't realize it was how JSONFormatter gets supplanted into the actual test ebin itself that it was trimming.

Adding `prune_code_paths: false` to the `mix.exs` of each test directory allows the tests to pass fully in `run-tests-in-docker.sh`

angelikatyborska commented:
I don't understand why the CI is still failing, adding the flag solved the issue  of missing results.json for me locally and when running `act -j smoke_test_in_docker`

dabaer commented:
Neat, I didn't know there was a command line flag to disable it as well!

Josè does mention runtime bugs can manifest in a release, but the project isn't run in a release so I'm not sure if there are any ramifications otherwise.

It's hard to see on the GitHub app but it says the run is cancelled immediately rather than any error

dabaer commented:
> I don't understand why the CI is still failing, adding the flag solved the issue of missing results.json for me locally and when running `act -j smoke_test_in_docker`

It's possible since you are testing it locally, because the docker container runs COPY on the directory that your un-pruned _build directory is making it's way in and not getting recompiled.

From what I can see `--no-prune-code-paths` is only applicable to the `mix compile` command, but adding it on all instances in `Dockerfile` and `run.sh` that run `mix compile` didn't work for me either.

I just pushed a commit but it's the original bandaid I made to get the tests to run in docker, with `prune_code_paths: false` added to the `mix.exs` files, it seems to work when I run it locally

angelikatyborska commented:
> with prune_code_paths: false added to the mix.exs files

I'm not sure if we can do that. Those mix projects in this repo are just test data and now the test data is different then the real data (exercise submissions from users). If adding `prune_code_paths: false` to `mix.exs` is really required, it has to happen as part of the test transformations done by the `exercism_test_helper`. Why? Well, if that change is required to make the test runner work, then every single exercise submission needs to have that change. But we cannot just modify the exercises in the source repo https://github.com/exercism/elixir because there are plenty of students that started solving those exercises already and the test runner would simply stop working for them.

angelikatyborska commented:
> It's possible since you are testing it locally, because the docker container runs COPY on the directory that your un-pruned _build directory is making it's way in and not getting recompiled.

You're right, after deleting all possible `_build` and `deps` directories, the problem came back when running `act -j smoke_test_in_docker`. I tried adding `--no-prune-code-paths` to all possible mix commands in the Dockerfile and all the scripts without success 😛 I wish I understood the compilation process better... but I don't think that this flag will bring us any luck.

Thankfully `mix.exs` is just another regular elixir file, so adding a transformation to it to add `prune_code_paths: false` to the list in the `project` function should be fairly straightforward with the usual metaprogramming techniques. Do you want to give it a try, or should I take over?

dabaer commented:
> > with prune_code_paths: false added to the mix.exs files
> 
> I'm not sure if we can do that. Those mix projects in this repo are just test data and now the test data is different then the real data (exercise submissions from users). If adding `prune_code_paths: false` to `mix.exs` is really required, it has to happen as part of the test transformations done by the `exercism_test_helper`. Why? Well, if that change is required to make the test runner work, then every single exercise submission needs to have that change. But we cannot just modify the exercises in the source repo https://github.com/exercism/elixir because there are plenty of students that started solving those exercises already and the test runner would simply stop working for them.
> 
> 

Yeah I considered it a temporary fix for testing, I figured we would need some sort of transformation for the real test data but wasn't sure if it were possible.

I'll give it a shot on adding those in the transformation step

dabaer commented:
@angelikatyborska Okay I've implemented the mix transformer to add the option to the test project mixfiles, but I'm concerned that at all points of these changes when running `bin/test_all_exercises.sh` I haven't seen that fail yet.

I haven't applied the mix transformer to that script, yet the tests all pass, so I'm not sure what is so different there that the transformation was succeeding all this time outside of the smoke tests.

angelikatyborska commented:
> EDIT: I guess it's due to the fact that it's not comparing the JSON results like we do in the smoke test.

Indeed, it's not even using the JSONFormatter and not loading the Jason library. I think the main intention behind `test_all_exercises` was ensuring that the test file transformations don't break the tests 🙂 

> If the addition of a command line argument for transforming the mix file in this fashion looks good to you, I will update the rest of the documentation in the repo around this new flag

Yes! I think this is the right approach and we're very close to merging this, finally 🙏 a documentation update is definitely necessary, maybe some comment in the new MixTransform module about the Elixir 1.15 changes too, as a hint to why we're doing all of this for our future selves.

neenjaw commented:
I'd still like to see the images updated if that is possible. Both the current and the proposed versions will soon not be supported. 

Otherwise looks good to me

angelikatyborska commented:
> I'd still like to see the images updated if that is possible. Both the current and the proposed Debian versions will soon not be supported.

Ah, I missed that, sorry! Yes, TL;DR: we want debian bookworm, or keep bullseye. Buster is older than bullseye and buster's security support ended in September 2022. Bullseye's security support ends in July 2024. Bookworm is very fresh, released in June 2023. Source: https://en.wikipedia.org/wiki/Debian_version_history

It's not something that I have to deal with anywhere else so I missed it 😅 it's easier with Ubuntu, just numbers.

dabaer commented:
> I'd still like to see the images updated if that is possible. Both the current and the proposed Debian versions will soon not be supported.
> 
> Otherwise looks good to me

Sorry about that it slipped my mind. I'll update to bookworm on all 3 repos as necessary

neenjaw commented:
> It's not something that I have to deal with anywhere else so I missed it 😅 it's easier with Ubuntu, just numbers.

Totally agree, Debian's versioning is harder to follow for this reason 😄 

> Sorry about that it slipped my mind. I'll update to bookworm on all 3 repos as necessary

No worries, thanks for the work done here!

---

Just for context, this whole pre-compilation and side-loading the compiled formatter binary was always a bit of a hack, even needing a bit of advice from @josevalim in the Elixir forum to get it past some compilation changes in 1.12 (ref missing). Nice work finding a way around it -- I wish we had a more standard way to do this as I'm sure we aren't the only language track with this sort of problem.

In the long term, it might be worth looking into expanding the mix transformer to inject the Jason dependency as either an included submodule or a [locally included dependency](https://hexdocs.pm/mix/Mix.Tasks.Deps.html) so that it could be compiled alongside the solution removing the need for side loading.  Thought that might have an adverse effect on the compilation/test time.

dabaer commented:
It's definitely a bit of dark magic to get this to work, but I love it.

I vaguely looked at trying to include the module as a dependency, and trying to see why a direct module inclusion would be pruned but the documentation around the pruning isn't quite there yet so I suppose we'll have to wait until it matures a bit.

A possibility would be rolling JSONFormatter into it's own hex package or similar, that way we can include it in `extra_applications` like we do for general dependencies that would otherwise be pruned, but that is just a shower thought.
